src/main/java/games/strategy/engine/data/Route.java

```diff
@@ -147,21 +147,16 @@ public Territory getStart() {
    */
   public boolean crossesWater() {
     final boolean startLand = !m_start.isWater();
-    boolean overWater = false;
-    final Iterator<Territory> routeIter = m_steps.iterator();
-    Territory terr = null;
-    while (routeIter.hasNext()) {
-      terr = routeIter.next();
-      if (terr.isWater()) {
-        overWater = true;
-      }
-    }
-    if (terr == null) {
+    final boolean overWater = m_steps.stream().anyMatch(Territory::isWater);
+
+    final Territory lastTerritory = m_steps.get(m_steps.size() - 1);
+
+    if (!overWater || lastTerritory == null) {
```

Comment:
Are we adding an extra constraint here by checking `!overWater`?  That doesn't appear to be in the original code.

Also, related to my previous comment, the original `terr == null` check was probably a test for if `m_steps` was empty.  I would assume we don't store `null` elements in `m_steps`.

Reply:
I wasn't sure about this, since we allow null values as start and end of a route (hat to deal with that when working on the route drawing code), but it seems the only way to add to this list is by calling Route#add which throws an exception if the passed territory is null, so you are indeed correct.

Reply:
Noting for an incompatible release:
We should merge the steps and the start territory to a single list, and change the getter methods accordingly for the sake of consistency